### ISSUE TITLE
cmp option parsing not like others

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#12747]
The code in apps/cmp.c has a bunch of " not like the others" code for parsing options.  For example, it scans argv specially to look for config file/section to do them first.  No other command has that.

It is not clear what the semantics of this behavior are.  Is it just to set up defaults? Other programs do that, why does this do it differently?

The cleverness with the union seems unwarranted.  Why can't this be done the way the s_client/server commands do thing?  What about writing a routine like the ssl parse config function?

**Comment by mattcaswell** [comment]
@DDvO

**Comment by DDvO** [comment]
CMP option parsing is not like others because its option handling is more mighty than others :smile:

This specific behavior of `apps/cmp.c` is explained briefly in the comment
```
     * handle OPT_CONFIG and OPT_SECTION upfront to take effect for other opts
```
User guidance on the `-config` and `-section` option can be found in the usual place: the man page.
See in particular the subjection `Using a custom configuration file` of the EXAMPLES.

The point is that a CMP client is more complex than (most) other OpenSSL apps because it typically needs a whole bunch of options.
Giving them on the command line is tedious and error-prone.
So I designed the CMP client to make heavy use of the configuration file mechanism, such that the user can specify most options statically in a config file of her choice and section(s) of her choice and simply refer to them e.g., via
```
openssl cmp -config myconfig.cnf -section mysection
```
The sections used may even be hierarchical, which is very handy when there are basic configurations per CMP server and use-case specific additions or overrides, to be used e.g., via
```
openssl cmp -config cmp.cnf -section server1,usecase3
```
The option values given in the config file section(s) serve as user-defined default values,
which in turn may be overridden and/or extended on-the-fly on the command line, e.g.,
```
openssl cmp -config cmp.cnf -section server1,usecase3 -certout my.crt
```

**Comment by DDvO** [comment]
BTW, in #12739 I had to extend the upfront treatment of CMP CLI options to the `-verbosity` that I added there
because it should take effect already for any diagnostic output during parsing of other options given in config files and/or on the same command line.

**Comment by DDvO** [comment]
BTW, IMHO the command line options of apps should be order-independent as far as possible.
Unfortunately there are a couple of OpenSSL apps that do not (fully) follow this principle,
such that things can go wrong unexpectedly just because the user entered options in 'wrong' order.
This sort of usability restriction is typically not documented for the respective apps.

**Comment by levitte** [comment]
I agree that a more stringent interpretation of options would be a good thing.

BTW, we already have this sort of thing, where some apps cache all the option values and act on them after parsing the command line, while others act immediately on the options as they are parsed. That is an inconsistency we have had for a loooong time.

**Comment by richsalz** [comment]
I don't expect anything to change for the 3.0 release.  The question is if it is worth having one command that is unlike the others.  As I initially said, s_server/s_client also read from a config file.

**Comment by DDvO** [comment]
> I don't expect anything to change for the 3.0 release. The question is if it is worth having one command that is unlike the others. As I initially said, s_server/s_client also read from a config file.

@richsalz, while the `s_server` and `s_client` (and BTW also `req`) apps can make some use of config files,
as I explained above they do not give the flexibility provided by the `cmp` app.
In particular, they do not allow extending and/or overriding configuration file options on the command line
and do not allow use of multiple sections where earlier (default) options can be extended/overridden by more specific ones.

A bettter question in my view would be which (other) apps should also offer such extended capabilities.

**Comment by richsalz** [comment]
> A bettter question in my view would be which (other) apps should also offer such extended capabilities.

Maybe that's a better question.  To me, a better question is if all that flexibility and code complexity for one command is needed.

**Comment by DDvO** [comment]
> To me, a better question is if all that flexibility and code complexity for one command is needed.

@richsalz ,I'm having the impression that you do not see the complexity of the CMP protocol and have no experience using it.
Otherwise you would appreciate the extended options support by the `cmp` app, strongly alleviating its use.

**Comment by DDvO** [comment]
Again, I see it the other way round: the `cmp` app pioneers powerful and convenient combined use of options in config files and on the command line. 
Other apps could benefit from this. To this end, the more general parts of its implementation could be moved to `apps/lib/opt.c` such that the required (not very heavy) code complexity would be just there.

**Comment by richsalz** [comment]
@ddvo, don't take offense.  This is about trade-offs and I happen to think this went the wrong way.  You don't, it got approved, and I'm not a member of the project so take this as external commentary.

I think "being a pioneer" has its downsides, as opposed to consistency.  Shrug.  Reasonable people may disagree.

If the cmp command is truly a pioneer, then leave this issue open until the wild west is settled :)

**Comment by nhorman** [comment]
so, 4 years later....There doesn't appear to be a consensus on what to do, and while its not great, the command line interpreters generally function as they should for their individual applets.

Marking as inactive, to be closed at the end of 3.4 dev barring further input.  We can open a new issue when we have a driving motivator to really address this

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
